server,sql,upgrade: add locality aware addresses for shared secondary tenants

shaikzakiriitm opened the pull request:
### Locality Aware Addresses for shared secondary tenants
As part of this change, we work on unifying the [initialization path for sqlInstanceDialer](https://github.com/cockroachdb/cockroach/blob/60831713192a69527ccd754220dd84a0bcb26456/pkg/server/server_sql.go#L608-L622) for both system and secondary tenants. Following are the list of steps (each step corresponds to a commit in this pr) done to achieve this change. 

- [ ] Add a jsonb column, `locality_address_list` to the system.sql_instances table. This contains the list of advertised addresses for given localities, of current sql instance. Say a source sql instance A needs to dail in target sql instance B. Instance row for B in sql_instances table will have values populated under this new column like "region=us-west@24.156.99.202:26257,zone=us-west-1b@172.217.22.14:26258" (actual data stored will be JSONified). These are the locality aware addresses which instance B is advertising for its callers to use. Let's say locaity of instance A is "locality=cloud=gce,region=us-west1,zone=us-west-1b". So, instance A, based on its own locality, will pick best address to dial instance B (in this case, `172.217.22.14:26258`). This matching of localities logic can be explained from [this snippet](https://github.com/cockroachdb/cockroach/blob/561a43d20ce870dbe0ad12fc3e51e3bee49c60e7/pkg/roachpb/metadata.go#L567). 
- [ ]  Updated instancestorage.go, instancereader.go, and added codec in row_codec.go for encoding/decoding instance row with the new column of locality address list. Here, instance row depicts a single row within the `system.sql_instances` table. As part of this commit, we update all the apis exposed to write/read instance row from `system.sql_instances` table by writing/reading the new column.
- [ ] For system tenants, we need to populate the locailty aware advertised addresses from the parent server (kv server) configurations. When we bring up the kv server on a node, we pass the flag ` --locality-advertise-addr localityList` as part of the `./cockroach start` command. System tenant inherits these values as its locality aware advertised addresses as both system tenant process and kv process are multiplexed on this address. This commit populates these values for the new tenant server created.
- [ ] When a sql server instance is brought up, especially the secondary tenants, OS auto-allocates ports to the sql instance process. These ports are different from the port which is already allocated to system tenant on that node. For secondary tenants, we need to populate these OS allocated ports as part of locality aware advertised address instead of the ports inherited from parent server configurations (i.e. instead of the ports provided with the flag ` --locality-advertise-addr localityList`). This commit updates the port populated for secondary tenants to those allocated by OS.
- [ ] Unify the code path of initialization of sqlInstanceDialer in the creation of sql server for both system and secondary tenants. 

**Note**: There are some build failures because of the issue called out in comment https://github.com/cockroachdb/cockroach/pull/143051#discussion_r2068638761. This is currently being worked upon, pls await further updates on the pr. 

Epic: CRDB-46923
Fixes: #138715
Part of: CRDB-46312

Release note: None

shaikzakiriitm committed:
sql,upgrade,clusterversion: add column locality_address_list

sql_instances table didn't have a way to store locality aware addresses for a sql instance
which gets bootstrapped.
Locality aware addresses will be used when dialing into a specific sql instance by the
sqlInstanceDialer.
This helps in routing the distributed sql query to the appropriate locality based address
of a selected sql instance.
As part of this change, added a column `locality_aware_address_list` of type jsonb into the
sql_instances table. Also added version gate, and a migration upgrade for adding this new column.
Included some unit tests around the same, testing addition of new column
in the table. Includes some updates due to rebase on latest master.

Epic: CRDB-46923
Fixes: #138715
Part of: CRDB-46312

Release note: None

cthumuluru-crdb commented:
Based on the comment [here](https://github.com/cockroachdb/cockroach/issues/90172#issuecomment-1284238333), this may not be immediately needed for external shared tenants. 

We can choose to defer these changes and track it using a backlog. I think it is also okay to commit these changes since the whole `mt` command is hidden and is not documents. I'm fine either way.

shaikzakiriitm committed:
sql: update codec for sql_instances table

sql_instances table didn't capture list of locality addresses that can be used to dial
the instance.
Needed to add this column in order to persist locality addresses for a sql_instance, in order to
unify sqlInstanceDialer initialization for both system and application tenants. With this
unification, we don't need to depend on gossip layer to identify locality addresses
to dial into system/shared tenants.
As part of this commit, added logic to encode/decode locality_address_list column in
sql_instances table. Also updated instancestorage.go to take in localityAddressList argument
in various methods to encode and store this value as part of the sql_instances table entry.
Added testing knob for unit tests to enable version guard while testing this new column. Updated
unit tests of instancestorage and instancereader.

Epic: CRDB-46923
Fixes: #138715
Part of: CRDB-46312

Release note: None

shaikzakiriitm committed:
server: propogating locality aware addresses to tenants

Locality aware addresses were not populated as part of the
system.sql_instances table prior to this change.
Locality aware addresses should be populated in the sql_instances
system table, as this information will be later used in the
sql instance dialer, by reading it with sqlInstanceReader.
As part of this commit, the locality aware addresses are being passed
on from the parent server configuration to that of the tenant sql server.

Epic: CRDB-46923
Fixes: #138715
Part of: CRDB-46312

Release note: None

shaikzakiriitm committed:
server: set secondary tenant address to have advertised port

Secondary shared tenant's address was stored as part of
system.sql_instances table. But the port used was that of
system tenant.

This needed to be updated because a secondary tenant should
not end up using the same port as that of system tenant.
System tenant acquires this port either from the command to
start the node or when OS allocates the port automatically.

To fix this, the port which is auto-allocated by OS to the
shared secondary tenant's sql process (stored/updated as part
of advertised address) is populated in system.sql_instances
table. Added unit tests for
updateLocalityAddressesForSharedSecondaryTenants function.

Epic: CRDB-46923
Fixes: #138715
Part of: CRDB-46312

Release note: None

shaikzakiriitm committed:
server: unify sqlInstanceDialer initialization

Previously, sqlInstanceDialer was instantiated differently
for system and secondary tenants. For system tenant,
kvNodeDialer was reused as sqlInstanceDialer. For secondary
tenants, a custom address resolver (using sqlInstanceReader)
was used to instantiate sqlInstanceDialer. kvNodeDialer used
locality aware addresses to identify the right ipaddress+port
to dial into the right node, but custom address resolver used
in sqlInstanceDialer of secondary tenants didn't have this
capability.

We needed to unify these initialization paths to reduce the
gap between system and secondary tenants as part of UA migration.

The locality aware addresses of sql instance are stored for
shared secondary tenants as part of sql_instances table. For
system tenant these values are obtained from the node's locality
aware addresses. For secondary tenants, their advertised address
with OS assigned ports are populated as part of locality aware
addresses. New custom address resolver is created which consults
the sql_instances table, as part of the sqlInstanceDialer.

Epic: CRDB-46923
Fixes: #138715
Part of: CRDB-46312

Release note: None

shaikzakiriitm commented on pkg/server/server_sql.go:
```diff
@@ -600,25 +602,38 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
-		// SQL pod addresses.
-		addressResolver := func(nodeID roachpb.NodeID) (net.Addr, roachpb.Locality, error) {
-			info, err := cfg.sqlInstanceReader.GetInstance(cfg.rpcContext.MasterCtx, base.SQLInstanceID(nodeID))
+	if /*true*/ cfg.Settings.Version.IsActive(ctx, clusterversion.V25_2) { // is this right key?
```
A bunch of tests in CI/CD pipeline are failing because of this condition. At this point, we notice that the cluster version is not yet created or instantiated. 
We need to explore other paths to get this value as part of the `newSQLServer` function. 
The cluster version is actually instantiated/initialized in the `preStart` method as can be seen [here](https://github.com/cockroachdb/cockroach/blob/145b2b717f132d8da38e4d5f51a2b5f6219dd64f/pkg/server/server_sql.go#L1469-L1517).

So, to use the cluster version check here, we are thinking of one alternative, where we can create a new custom addressResolver and have the if/else check within it, where if uses the addressResolver logic of kvNodeDialer (gossip based resolver) and the else uses this address resolver which matches locality address, although this digresses from general philosophy to have the fallback logic of this feature flag (version gate) identical to the existing code.

shaikzakiriitm replied:
The other alternative is to wrap the instantiation/initialization of `cfg.sqlInstanceDialer` within a function and invoke this function in the `preStart` method after the cluster version is initialized. Although, we need to test it out if there are not scenarios where sqlInstanceDialer is used after instantiation of `newSQLServer` but before `preStart` on this instance is invoked. This still digresses from the fallback logic of feature-flag to look as identical as it is today.

shaikzakiriitm replied:
@cthumuluru-crdb @stevendanna any further thoughts on this? Thanks!

stevendanna commented on pkg/server/server_sql.go:
```diff
@@ -600,26 +602,39 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
 			}
-			return &util.UnresolvedAddr{AddressField: info.InstanceRPCAddr}, info.Locality, nil
+			defaultAddress := &util.UnresolvedAddr{AddressField: info.InstanceRPCAddr}
+			return cfg.Locality.LookupAddress(info.LocalityAddressList, defaultAddress), info.Locality, nil
```
Could we unify this with the sql instance resolve we use below? If LocalityAddressList is empty this function is a no-op so we should be able to use the same implementation for both in either case.  I understand the desire to keep the old code path unchanged, but that is really only valuable if we also have a cluster setting that lets you opt-opt of it.

stevendanna commented on pkg/server/server_sql.go:
```diff
@@ -1792,6 +1814,34 @@ func (s *SQLServer) preStart(
 	return nil
 }
 
+// updateLocalityAddressesForSharedSecondaryTenants updates the locality addresses of this sql instance
+// if the instance is a shared secondary tenant. We update the port part of each address to that of the
+// advertised address as this will be stored in the sql_instances table and will be consulted by the
+// sqlInstanceDialer when attempting to dial this sql instance.
```
I think what I want to know from this comment is why locality addresses aren't correct by default.

shaikzakiriitm replied:
In short, have removed this function now as its not needed.

stevendanna reviewed (commented):
Thanks for digging into this.

I gave this a first pass.

shaikzakiriitm commented on pkg/sql/catalog/systemschema/system.go:
```diff
@@ -3944,13 +3946,14 @@ var (
 					{Name: "crdb_region", ID: 6, Type: types.Bytes, Nullable: false},
 					{Name: "binary_version", ID: 7, Type: types.String, Nullable: true},
 					{Name: "is_draining", ID: 8, Type: types.Bool, Nullable: true},
+					{Name: "locality_address_list", ID: 9, Type: types.Jsonb, Nullable: true},
```
Yes I thought of both the names, `locality_address_list` as well as `locality_addresses` while adding this column. I wanted to pick a name which is more indicative of the fact that underlying json datum is a list of locality addresses. 
The choice here is also inspired from the fact that the documentation in the cli config uses `list` to describe this config. 
```
      --locality-advertise-addr localityList
                                                        List of ports to advertise to other CockroachDB nodes for intra-cluster
                                                        communication for some locality. This should be specified as a comma separated
                                                        list of locality@address. Addresses can also include ports. For example:

                                                          "region=us-west@127.0.0.1,zone=us-west-1b@127.0.0.1"
                                                          "region=us-west@127.0.0.1:26257,zone=us-west-1b@127.0.0.1:26258"
```
Happy to stick with `locality_address_list`, but open to renaming it if you have a strong preference for an alternative. Thanks!

shubhamdhama replied:
I'd still suggest using LocalityAddresses. The `List` suffix seems redundant. Using `LocalityAddress` would also keep things more consistent with other parts of the codebase, https://github.com/cockroachdb/cockroach/blob/11c36656ac659a4965e0a290cc9c4a4fadec3bce/pkg/server/serverpb/status.proto#L181-L200

https://github.com/cockroachdb/cockroach/blob/e00a5ad419b001824f2309b06fe926d065c196d4/pkg/base/config.go#L475-L477

https://github.com/search?q=repo%3Acockroachdb%2Fcockroach%20LocalityAddresses&type=code

Yeah, for the CLI interface, list makes sense. But that didn't really change how we named the field in `NodeDescriptor` before, so I figure we can stick to the consistent naming now too. It's your call in the end, though. I'm good either way.

shaikzakiriitm commented on pkg/server/server_sql.go:
```diff
@@ -1572,24 +1564,9 @@ func (s *SQLServer) preStart(
-					s.execCfg.Settings.Version.LatestVersion(),
-					nodeID,
-				)
+				return s.sqlInstanceStorage.CreateNodeInstance(ctx, session, s.cfg.AdvertiseAddr, s.cfg.SQLAdvertiseAddr, s.distSQLServer.Locality, s.execCfg.Settings.Version.LatestVersion(), nodeID, []roachpb.LocalityAddress{})
```
done 👍

shaikzakiriitm committed:
fixup! server: unify sqlInstanceDialer initialization

shaikzakiriitm committed:
fixup! sql,upgrade,clusterversion: add column locality_address_list

shaikzakiriitm committed:
fixup! server: unify sqlInstanceDialer initialization

shaikzakiriitm committed:
fixup! sql: update codec for sql_instances table

shaikzakiriitm committed:
fixup! sql,upgrade,clusterversion: add column locality_address_list

shubhamdhama commented:
This is looking great!

A few nitpicks:

Please take a look at the [git commit guidelines](https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages). I find they're pretty applicable to PR descriptions as well. You might want to remove the checkboxes, as they could make it seem like some tasks are still pending.

For the title, consider dropping `upgrade` from the prefixes. Something like `sql, server` could work. 

Also, there are few typos in the PR description.

cthumuluru-crdb reviewed (changes requested):
Overall, changes look good. I left a few comments. Please take a look. Thanks!

cthumuluru-crdb commented on pkg/server/server_sql.go:
```diff
@@ -600,26 +601,28 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
+			return resolveWithSQLInstanceReader()
 		}
-		cfg.sqlInstanceDialer = nodedialer.New(cfg.rpcContext, addressResolver)
+		g, _ := cfg.gossip.Optional(138715)
```
Well, I'm wrong about this. Please ignore my comment. No I remember we really want `cfg.Settings.Version.IsActive(ctx, clusterversion.V25_3)` check to be deferred and so it had to be inside address resolver.

cthumuluru-crdb replied:
How about passing `kvAddressResolver` or `gossipAddressResolver` to SQL server and using it here? 
<!-- Reviewable comment -OQn_dyA1-nh3piaWksN:b-ac2hev -->
<!-- Sent from Reviewable.io -->

cthumuluru-crdb reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143051)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jeffswenson, @mw5h, @shaikzakiriitm, @shubhamdhama, and @stevendanna)



<!-- Sent from Reviewable.io -->

mw5h commented:
No activity on this PR in over a year, okay to close it out?

shaikzakiriitm commented:
> No activity on this PR in over a year, okay to close it out?

Hi @mw5h , I'm not actively working on this pull request. Feel free to discuss with the team and close it. Thanks!
